feat: syscall phase1 구현

hdh853 opened the pull request:
## 🔗 관련 이슈
<!-- 이슈 번호를 연결해주세요 -->
closes #23 

## 📌 작업 내용
### 구현 사항
<!-- 구현내용에 대해서 설명해주세요. -->
<!-- 예시 :
- 이슈 템플릿, PR 템플릿 작성
- READEME 파일 최초 작성
 --> 
- syscall phase_1 구현
  - create()
  - remove()
  - halt()
  - open()
  - close()
  - filesize()
  - seek()
  - tell()
  - read()
  - write()
  - exit()


### 변경된 파일
<!-- 구현 내용으로 인해 변경된 파일에 대해 설명해주세요 -->
<!-- 예시:
- README.md
  - readme 파일 최초 생성
- issue_template.md
  - 이슈 템플릿 최초 생성
-->
- syscall.c
  - phase_1 시스템콜 추가
  - get_file_from_fd, check_valid_addr 헬퍼함수 추가
- thread.h
  -  exit 구현을 위한 exit_code
  -  fd테이블 관리를 위한 open_file_list 멤버 추가
  
- thread.c
  -  쓰레드 name 인자 제외하고 설정되도록 수정
  - open_file_list 초기화 추가
- process.c
  - exit 출력문 추가 

### 추가 사항
<!-- 트러블 슈팅 등 추가로 설명할 내용이 있으시다면 아래에 작성해주세요. -->
- 예전에는 배열로 fd테이블을 구현했는데, fd테이블 크기에 제한을 두지 않기 위해 list로 관리하려 하면서 list다루는데 시간 소요

## ✅ 셀프 체크리스트
<!-- 체크리스트를 모두 완료한 후 제출해주세요 -->
<!-- 완료 처리를 표시하려면 [x], 완료되지 않았다면 [ ] -->
- [x] 코드 컨벤션을 준수했나요?
- [x] 불필요한 주석이나 디버깅 코드를 제거했나요?
- [x] 관련 테스트가 모두 통과하나요?
- [x] 커밋 메시지가 컨벤션에 맞나요?

hdh853 committed:
style: syscall.c 템플릿 적용

hdh853 committed:
feat: syscall_phase1 구현
- create()
- remove()
- halt()
- open()
- close()
- filesize()
- seek()
- tell()
- read()
- write()
- exit() - thread.c에서 출력문만 구현

hyunmuam commented on userprog/process.c:
```diff
@@ -202,6 +202,7 @@ int process_wait(tid_t child_tid UNUSED)
     /* XXX: Hint) The pintos exit if process_wait (initd), we recommend you
      * XXX:       to add infinite loop here before
      * XXX:       implementing the process_wait. */
+    timer_sleep(100);
```
디버깅용 코드 같은데 제거해주시면 좋을 것 같습니다.

hdh853 replied:
놓쳤네요. 지우도록 하겠습니다.

hyunmuam commented on userprog/process.c:
```diff
@@ -214,6 +215,9 @@ void process_exit(void)
      * TODO: We recommend you to implement process resource cleanup here. */
 
+    // process termination message
+    printf("%s: exit(%d)\n", curr->name, curr->exit_code);
```
> 사용자 프로세스가 아닌 커널 스레드가 종료되거나 halt 시스템 호출이 호출될 때는 이러한 메시지를 출력하지 마십시오.

스펙에서는 커널 스레드 종료 시 출력하지 말라고 명시되어 있는데, 현재 코드만 보았을 경우 모든 경우에 출력되는 것 같습니다. 현재 어떻게 구분하고 계신가요?

hdh853 replied:
항상 출력되도록 구현을 했네요.. 조건 설정해서 유저프로그램인 경우에만 출력하도록 수정하겠습니다

hyunmuam commented on userprog/syscall.c:
```diff
@@ -24,23 +28,294 @@ void syscall_handler (struct intr_frame *);
+void syscall_handler(struct intr_frame* f)
+{
+    // 인자 순서 : rdi, rsi, rdx
+    // 시스템콜 번호 / 리턴 : rax
```
개발 과정에서만 필요한 임시 주석으로 보입니다. 
최종 코드에는 불필요한 것 같습니다. 제거 부탁드립니다.

hdh853 replied:
@tjvm0877 
다음 phase_2 개발할때 필요할것 같아서 남겨뒀었는데, 최종 코드에는 없는게 맞는것 같네요. 이럴 경우 phase_1 끝났을 때 미리 지워야 하나요 아니면 phase_2 끝났을 때 지워야 할까요?

hyunmuam replied:
@hdh853 
develop 브랜치는 프로덕션 수준의 코드만 머지되어야 한다고 생각해요.
개발 과정의 임시 주석은 제거 하는 것이 맞는 것 같습니다.

해당 주석이 향후 작업에 필요하다면 phase2 브랜치를 생성할 때 
다시 추가하는 방식으로 진행하면 될 것 같습니다.

hyunmuam commented on userprog/syscall.c:
```diff
@@ -24,23 +28,294 @@ void syscall_handler (struct intr_frame *);
+    case SYS_EXIT:
+        int status = f->R.rdi;
+        sys_exit(status);
+        break;
```
시스템 콜 함수를 호출할 때 아래와 같이 인라인으로 하지 않는 이유를 말씀해주실 수 있나요?

```c
sys_exit(f->R.rdi);
```

hdh853 replied:
인라인으로 해야겠다는 생각을 못했던게 제일 큰 이유인것 같습니다. 인라인으로 하는게 코드가 더 깔끔할 것 같네요

hyunmuam commented on userprog/syscall.c:
```diff
@@ -24,23 +28,294 @@ void syscall_handler (struct intr_frame *);
+    thread_exit();
+}
+
+static bool sys_create(const char* file, unsigned initial_size)
+{
+    check_valid_addr(file);
+
+    lock_acquire(&file_lock);
+    bool is_created = filesys_create(file, initial_size);
+    lock_release(&file_lock);
+    return is_created;
+}
```
`sys_create`, `sys_remove`, `sys_open`에서 아래와 같이 유저 주소 여부를 검사하는 것 같습니다. 
> `check_valid_addr(file);`

file은 문자열로 받는데 현재 방법으로 하면 문자열의 첫 번째 char의 유저 주소 여부만 검사하고 끝나버립니다.
문자열 전체를 검사하는 별도 함수가 필요할 것 같습니다.

hdh853 replied:
@tjvm0877 
문자열 전체를 검사해야 한다는 말씀이, '길이가 지나치게 크지 않은지', '널문자가 없는 문자열인지' 를 검사해야한다는 말씀이신거죠? 그부분은 추가해야할것 같네요.

hyunmuam replied:
@hdh853 
문자열은 char 배열이고 `file`은 첫 번째 문자를 가리킵니다.
현재 코드는 첫 번째 문자만 검증하므로, 악의적 사용자가 
문자열 중간에 커널 주소를 포함시키면 보안 취약점이나 페이지 폴트가 발생할 수 있습니다.

따라서 NULL terminator(`\0`)까지 모든 바이트를 검증해야 한다고 생각합니다.

hyunmuam commented on userprog/syscall.c:
```diff
@@ -24,23 +28,294 @@ void syscall_handler (struct intr_frame *);
+static int sys_read(int fd, void* buffer, unsigned size)
+{
+    check_valid_addr(buffer);
+    check_valid_addr(buffer + size - 1);
```
`sys_write`와 `sys_read`가 현재 버퍼의 시작과 끝만 검사하고 있습니다. 

버퍼가 여러 페이지에 걸쳐 있고 중간 페이지가 unmapped되어 있으면 page fault가 발생할 수 있다고 생각합니다. 이를 방지하기 위해 페이지 경계마다 검사하는 것이 더 안전할 것 같은데, 이 부분 어떻게 생각하시나요?

hdh853 replied:
그러네요. 중간에 unmapped 될 수도 있는 케이스를 고려하지 못했네요. 검사하는게 맞을것 같습니다

hyunmuam commented on userprog/syscall.c:
```diff
@@ -24,23 +28,294 @@ void syscall_handler (struct intr_frame *);
+
+static void sys_close(int fd)
+{
+
+    // open_file_list 에서 제거
+    struct list* open_file_list = &thread_current()->open_file_list;
+    struct list_elem* e;
+
+    for (e = list_begin(open_file_list); e != list_end(open_file_list); e = list_next(e)) {
+        struct open_file_list_elem* fd_entry = list_entry(e, struct open_file_list_elem, elem);
+        if (fd_entry->fd == fd) {
+            list_remove(&fd_entry->elem);
+
+            lock_acquire(&file_lock);
+            file_close(fd_entry->file);
+            lock_release(&file_lock);
+            break;
+        }
+    }
+}
+
+// fd로 file 찾는 함수
+static struct file* get_file_from_fd(int fd)
+{
+    struct list* open_file_list = &thread_current()->open_file_list;
+    struct list_elem* e;
+
+    for (e = list_begin(open_file_list); e != list_end(open_file_list); e = list_next(e)) {
+        struct open_file_list_elem* fd_entry = list_entry(e, struct open_file_list_elem, elem);
+        if (fd_entry->fd == fd) {
+            return fd_entry->file;
+        }
+    }
+    return NULL;
+}
```
두 함수에서 순회하는 부분을 공통화시킬 수 있을 것 같습니다. 

예를 들어 `find_fd_elem(int fd)` 같은 헬퍼 함수를 만들어 
`open_file_list_elem*`을 반환하도록 하면 코드 중복을 줄일 수 있을 것 같습니다.

hdh853 replied:
그러네요. fd 기준으로 open_file_list_elem 구조체 리턴받는 공통함수 빼고, 그걸 가져와서 각자 처리를 다르게 하는 방식으로도 하면 되겠네요 감사합니다

hyunmuam commented on threads/thread.c:
```diff
@@ -419,11 +419,14 @@ static void init_thread(struct thread* t, const char* name, int priority)
     memset(t, 0, sizeof *t);
     t->status = THREAD_BLOCKED;
     strlcpy(t->name, name, sizeof t->name);
+    char* savePtr;
+    strtok_r(t->name, " ", &savePtr);
     t->tf.rsp = (uint64_t)t + PGSIZE - sizeof(void*);
     t->priority = priority;
     t->magic = THREAD_MAGIC;
     t->original_priority = priority;
     list_init(&t->donations);
+    list_init(&t->open_file_list);
 }
```
```c
strtok_r(t->name, " ", &savePtr);
t->tf.rsp = (uint64_t)t + PGSIZE - sizeof(void*);
```
간단하게 이름을 파싱하셨네요!

다만 `argument_parse`에서 이미 파싱된 이름을 넘겨주기 때문에 
여기서 다시 파싱하는 것은 중복 연산인 것 같습니다.
문제가 발생하는 이유인 `process_create_initd`에서 파싱된 이름을 받아서 설정하는 방식으로 변경하면 어떨까요?

hdh853 replied:
그러네요 process_create_initd에서 thread_create()를 하기전 파일명만 잘라서 넘겨주면 유저프로세스 일때만 파일명을 잘라서 주는게 가능하겠네요. 수정하도록 하겠습니다!

hdh853 committed:
refactor: 피드백 반영
 - 불필요한 주석 제거
 - syscall.c 유효성 검사 로직 추가
 - thread.c에서 파일명 처리 로직 삭제
 - process.c 에서 파일명 처리 로직 추가
 - process.c 에서 exit 조건문 추가

hyunmuam commented on userprog/syscall.c:
```diff
@@ -24,23 +28,317 @@ void syscall_handler (struct intr_frame *);
+        return -1;
+    }
+    struct thread* curr = thread_current();
+    int fd = list_size(&curr->open_file_list) + 2; // 0과 1은 표준입,출력
```
@hdh853 
Size로 계산을 때려버리면 동일한 FD가 생길 수 있을 것 같습니다.

hdh853 replied:
그러네요 로직을 바꿔보겠습니다

hyunmuam commented on userprog/syscall.c:
```diff
@@ -24,23 +28,317 @@ void syscall_handler (struct intr_frame *);
+{
+    lock_acquire(&file_lock);
+
+    struct file* file = get_list_elem_from_fd(fd)->file;
```
@hdh853 
임계영역 지정의 일관성이 부족한 것 같습니다.

hdh853 replied:
넵 수정하겠습니다

hdh853 committed:
refactor: 피드백 반영
- sys_filesize 임계영역 수정
- sys_open fd값 부여 로직 수정

hdh853 committed:
refactor: 피드백 반영
- sys_open fd 부여 로직 수정

hyunmuam commented on userprog/syscall.c:
```diff
@@ -158,27 +165,48 @@ static int sys_open(const char* file)
+            list_insert(e->next, &fd_entry->elem);
+            return fd_entry->fd;
+        }
+    }
```
이런식으로 주석을 추가하면 더 코드를 이해하기 편할 것 같습니다.
```c
for (e = list_begin(open_file_list);
     e != list_end(open_file_list);
     e = list_next(e)) {

    struct open_file_list_elem *curr_entry =
        list_entry(e, struct open_file_list_elem, elem);

    /* 현재 노드가 마지막 실제 노드라면(다음이 tail 센티넬),
       마지막 fd 바로 다음 번호를 새로 할당하고 맨 뒤에 추가한다. */
    if (e->next == list_tail(open_file_list)) {
        fd_entry->fd = curr_entry->fd + 1;
        list_push_back(open_file_list, &fd_entry->elem);
        return fd_entry->fd;
    }

    /* 다음 노드가 있고,
       - 만약 다음 노드의 fd가 (현재 fd + 1)이면: 중간에 빈 fd가 없으므로 계속 순회한다.
       - 그렇지 않다면: (현재 fd + 1)이 비어 있는 최소 fd이므로, 그 자리에 새 항목을 끼워 넣는다. */
    struct open_file_list_elem *next_entry =
        list_entry(e->next, struct open_file_list_elem, elem);
    if (curr_entry->fd + 1 == next_entry->fd) {
        continue;
    } else {
        fd_entry->fd = curr_entry->fd + 1;
        list_insert(e->next, &fd_entry->elem);
        return fd_entry->fd;
    }
}
```

hdh853 replied:
그러네요 반영하도록 하겠습니다

hdh853 committed:
refactor: 피드백 반영
- 매직넘버 제거
- thread.h 멤버 위치 조정

hdh853 committed:
chore: sys_open 시스템콜 주석 추가

hyunmuam reviewed (changes requested):
`threads`빌드 실패했습니다. CI한번 확인해주세요.

hdh853 committed:
fix: init_thread시 fd list는 USERPROG일 경우에만 init 하도록 수정

hyunmuam reviewed (approved):
수고하셨습니다. 👍
